rabbit-escape-render/src/rabbitescape/render/WaterRegionRenderer.java

```diff
@@ -0,0 +1,364 @@
+            if ( null == wrr )
+            {
+                return height;
+            }
+            return wrr.height;
+        }
+        return height;
+    }
+
+    private int calcBoundaryHeight( CellularDirection d )
+    {
+        if ( !region.isConnected( d ) )
+        { // The cell in that direction is not relevant
+            return height;
+        }
+        WaterRegionRenderer adjWrr = adjacentRenderer( d );
+        if ( null == adjWrr )
+        { // Adjacent is probably empty, and this cell is probably a low level
+            return height;
+        }
+        int heightDefect = Math.abs( targetWaterHeight - height );
+        int adjHeightDefect = Math.abs ( adjWrr.targetWaterHeight - adjWrr.height );
+        if ( heightDefect > adjHeightDefect )
+        {
+            if ( heightDefect > maxHeightChange )
+            {
+                return height;
+            }
+        }
+        else
+        {
+            if ( adjHeightDefect > maxHeightChange )
+            {
+                return adjWrr.height;
+            }
+        }
+        int adjHeight = adjacentRendererWaterHeight( d );
+        int boundaryHeight = adjHeight > height ?  adjHeight : height; // Use max
+        return boundaryHeight;
+    }
+
+    public void bottomVertex( Vector<Integer> retX, Vector<Integer> retY, CellularDirection d )
+    {
+        int x = region.x * 32, y = region.y * 32; // Local cell origin in nominal pixels.
+        int xOffset;
+        Block block = world.getBlockAt( region.x, region.y );
+        switch ( d )
+        {
+        case LEFT:
+            if ( BehaviourTools.isLeftRiseSlope( block ) )
+            {
+                xOffset = 32;
+            }
+            else
+            {
+                xOffset = 0;
+            }
+            break;
+        case RIGHT:
+            if ( BehaviourTools.isRightRiseSlope( block ) )
+            {
+                xOffset = 0;
+            }
+            else
+            {
+                xOffset = 32;
+            }
+            break;
+        default:
+            return;
+        }
+        retX.add( x + xOffset );
+        retY.add( y + 32 );
+    }
+
+    boolean isFalling()
+    {
+        if ( BehaviourTools.isSlopeNotBridge( world.getBlockAt( region.x, region.y ) ) )
+        { // Non bridge slopes are not connected down.
+            return false;
+        }
+
+        return !isFull( region.x, region.y + 1 );
+    }
+
+    boolean isFull( int x, int y )
+    {
+        Block b = world.getBlockAt( x, y );
+        if ( BehaviourTools.s_isFlat( b ) )
+        { // Flat block cells have zero capacity: always full.
+            return true;
+        }
+        WaterRegionRenderer wrr = dynamics.lookupRenderer.getItemAt( x, y );
+        if ( null == wrr )
+        { // Not full of flat block, but no region yet, empty
+            return false;
+        }
+        return 32 == wrr.height && 32 == wrr.targetWaterHeight;
```

Comment:
the height calc takes account of slope blocks. so they are full when at 32, like empty cells.

targetWaterHeight tracks the contents of the WaterRegion from engine. So if a non-block cell is half full (content 512), the targetWaterHeight will be 16. the height lags behind a little to make the height changes look smoother.